near-plugins-derive/src/upgradable.rs

```diff
@@ -18,29 +20,98 @@ pub fn derive_upgradable(input: TokenStream) -> TokenStream {
+                    ::near_sdk::Duration::try_from_slice(&duration_bytes).unwrap_or_else(|_|
+                            near_sdk::env::panic_str("Upgradable: Invalid u64 Duration format")
+                    )
+                })
+            }
+
+            fn up_set_timestamp(&self, key: __UpgradableStorageKey, value: ::near_sdk::Timestamp) {
+                self.up_storage_write(key, &value.try_to_vec().unwrap());
+            }
+
+            fn up_set_duration(&self, key: __UpgradableStorageKey, value: ::near_sdk::Duration) {
+                self.up_storage_write(key, &value.try_to_vec().unwrap());
+            }
+
+            fn up_storage_key(&self, key: __UpgradableStorageKey) -> Vec<u8> {
+                let key_vec = key
+                    .try_to_vec()
+                    .unwrap_or_else(|_| ::near_sdk::env::panic_str("Storage key should be serializable"));
+                [(#storage_prefix).as_bytes(), key_vec.as_slice()].concat()
+            }
+
+            fn up_storage_write(&self, key: __UpgradableStorageKey, value: &[u8]) {
+                near_sdk::env::storage_write(self.up_storage_key(key).as_ref(), &value);
+            }
+
+            fn up_set_staging_duration_unchecked(&self, staging_duration: near_sdk::Duration) {
+                self.up_storage_write(__UpgradableStorageKey::StagingDuration, &staging_duration.try_to_vec().unwrap());
+            }
+        }
+
         #[near_bindgen]
         impl Upgradable for #ident {
-            fn up_storage_key(&self) -> Vec<u8>{
-                (#code_storage_key).as_bytes().to_vec()
+            fn up_storage_prefix(&self) -> &'static [u8] {
+                (#storage_prefix).as_bytes()
+            }
+
+            fn up_get_delay_status(&self) -> #cratename::UpgradableDurationStatus {
+                near_plugins::UpgradableDurationStatus {
+                    staging_duration: self.up_get_duration(__UpgradableStorageKey::StagingDuration),
+                    staging_timestamp: self.up_get_timestamp(__UpgradableStorageKey::StagingTimestamp),
+                    update_staging_duration: self.up_get_duration(__UpgradableStorageKey::UpdateStagingDuration),
+                }
             }
 
             #[#cratename::only(owner)]
             fn up_stage_code(&mut self, #[serializer(borsh)] code: Vec<u8>) {
+                near_sdk::require!(
+                    self.up_get_duration(__UpgradableStorageKey::UpdateStagingDuration).is_none(),
+                    "Upgradable: an update of staging duration is in progress"
+                );
+
+                let timestamp = near_sdk::env::block_timestamp() + self.up_get_duration(__UpgradableStorageKey::StagingDuration).unwrap_or(0);
+
                 if code.is_empty() {
-                    near_sdk::env::storage_remove(self.up_storage_key().as_ref());
+                    near_sdk::env::storage_remove(self.up_storage_key(__UpgradableStorageKey::Code).as_ref());
```

Comment:
Despite removing staged code a new `StagingTimestamp` is set below. Shouldn't removing staged code also remove the `StagingTimestamp`?

Reply:
The same as for staged code, it will be just an extra operation without any benefits. but it also could be useful to see the last timestamp as historical info.

Reply:
When removing staged code, wouldn't it be more intuitive to remove `StagedTimestamp` instead of setting it to a new value?

Reply:
Fixed 5c4fce0b5e25670a1d3810a00a08cfd566655486